### ISSUE TITLE
docker cp archive/tar error with filenames greater than 100 characters

### ISSUE THREAD

**Issue by mkozell** [docker/for-linux#484]
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior

Using `docker cp` should copy files and directories from container to host with filenames greater than 100 characters for Docker version **18.06.1-ce** and higher.

### Actual behavior

Using `docker cp container_name:/etc/resolv.conf a_very_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_long_filename_that_is_101_characters` in **18.03.1-ce** works but results in the error below on **18.06.1-ce** and higher.

```
archive/tar: cannot encode header: Format specifies USTAR; and USTAR cannot encode Name="a_very_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_long_filename_that_is_101_characters"
```

### Steps to reproduce the behavior

**Output of `docker version`:**
```
Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  
Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**

I was able to reproduce the issue in Docker **18.06.1-ce** and **18.09.0-ce** but not in Docker **18.03.1-ce**. The commands below can reproduce the issue. I reproduced the issue using CentOS 7.5 and Amazon Linux 2 hosts. I then used Vagrant to confirm the issue locally.

```
docker pull centos:7
docker run -d -i --name docker_test centos:7
docker cp docker_test:/etc/resolv.conf a_very_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_long_filename_that_is_100_characters
docker cp docker_test:/etc/resolv.conf a_very_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_long_filename_that_is_101_characters
```

It appears a change to Golang v1.10 may have changed the default archive/tar format to USTAR which has a 100 character Linkname limit.

https://golang.org/doc/go1.10#archive/tar
https://golang.org/pkg/archive/tar/#Format

**Comment by thaJeztah** [comment]
ping @kolyshkin

**Comment by travisreed-wf** [comment]
I am running into this issue in `18.06.1-ce` as well. I will downgrade to `18.03.1-ce` but wanted to chime in that it is impacting us too.

**Comment by kolyshkin** [comment]
The problem is, docker daemon writes tar in PAX format, but since the reading side has to guess the format, and it mis-guesses it as USTAR. This issue is probably caused in `archive/tar` changes in Go 1.10.

I have a fix in progress...

**Comment by kolyshkin** [comment]
Proposed fix: https://github.com/moby/moby/pull/38634 (note it needs to be merged to moby/moby first and then vendored to docker/cli in order to take effect).

**Comment by interactwithankush** [comment]
Assuming this issue is fixed in later docker version, could someone please share the version which solves this issue.

**Comment by thaJeztah** [comment]
Yes this is fixed;

```console
$ docker run -d -i  --name docker_test alpine
$ docker cp docker_test:/etc/resolv.conf a_very_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_long_filename_that_is_100_characters
$ docker cp docker_test:/etc/resolv.conf a_very_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_long_filename_that_is_101_characters
$ ls a_very_*
a_very_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_long_filename_that_is_100_characters
a_very_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_long_filename_that_is_101_characters
```

> could someone please share the version which solves this issue.

It's included in Docker 18.09.4 and up; https://github.com/docker/engine/pull/167 / https://github.com/docker/cli/pull/1778

**Comment by michelle-chou25** [comment]
Unfortunatly I meet this issue in a newer docker version:
Client: Docker Engine - Community
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:45:41 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:03:33 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

I run the cp command:
 docker cp consolidated.00.pth test_alpaca:/home/model_zoo/llama/7B
ERRO[0000] Can't add file /home/michellezhou/llama/7B/consolidated.00.pth to tar: archive/tar: write too long
